packages/web/queries/complex/pools.ts

```diff
@@ -0,0 +1,325 @@
+import { CoinPrimitive } from "@keplr-wallet/stores";
+import { Dec, DecUtils } from "@keplr-wallet/unit";
+import {
+  ConcentratedLiquidityPoolRaw,
+  CosmwasmPoolRaw,
+  StablePoolRaw,
+  WeightedPoolRaw,
+} from "@osmosis-labs/pools";
+import { CacheEntry, cachified } from "cachified";
+import { LRUCache } from "lru-cache";
+
+import { queryBalances } from "~/queries/cosmos";
+import {
+  FilteredPoolsResponse,
+  PoolToken,
+  queryFilteredPools,
+} from "~/queries/indexer";
+import { queryNumPools, queryPools } from "~/queries/osmosis";
+
+export type PoolRaw =
+  | CosmwasmPoolRaw
+  | StablePoolRaw
+  | ConcentratedLiquidityPoolRaw
+  | WeightedPoolRaw;
+
+export async function queryPaginatedPools(
+  page?: number,
+  limit?: number,
+  poolIdParam?: string
+): Promise<{ status: number; pools: PoolRaw[] }> {
+  // Fetch the pools data from your database or other source
+  // This is just a placeholder, replace it with your actual data fetching logic
+  const allPools: PoolRaw[] = await fetchAndProcessAllPools();
+
+  // Handle the case where specific pool ID is requested
+  if (poolIdParam) {
+    const pool = allPools.find(
+      (pool) => ("pool_id" in pool ? pool.pool_id : pool.id) === poolIdParam
+    );
+    if (!pool) {
+      throw { status: 404, pools: [] };
+    }
+    return { status: 200, pools: [pool] };
+  }
+
+  // Pagination
+  if (page && limit) {
+    const startIndex = (page - 1) * limit;
+
+    // Slice the data based on the page and limit
+    const pools = allPools.slice(startIndex, startIndex + limit);
+
+    // Return the paginated data
+    return { status: 200, pools };
+  }
+
+  // Return the paginated data
+  return { status: 200, pools: allPools };
+}
+
+/** Cache on this current edge function instance. */
+const allPoolsLruCache = new LRUCache<string, CacheEntry>({
+  max: 2,
+});
+
+async function fetchAndProcessAllPools(): Promise<PoolRaw[]> {
+  return cachified({
+    key: "all-pools",
+    cache: allPoolsLruCache,
+    async getFreshValue() {
+      const numPools = await queryNumPools();
+
+      // Fetch all pools from imperator, except cosmwasm pools for now
+      // TODO remove when indexer returns cosmwasm pools
+      try {
+        const filteredPoolsResponse = await queryFilteredPools(
+          {
+            min_liquidity: 0,
+            order_by: "desc",
+            order_key: "liquidity",
+          },
+          { offset: 0, limit: Number(numPools.num_pools) }
+        );
+        const queryPoolRawPromises = filteredPoolsResponse.pools.map(
+          queryPoolRawFromFilteredPool
+        );
+        const queryPoolRawResults = await Promise.all(queryPoolRawPromises);
+
+        const cosmwasmPools = await getCosmwasmPools();
```

Comment:
Since we don't have any dependencies coming from `queryFilteredPools`, we can place `getCosmwasmPools` and `queryFilteredPools` in a promise.all to execute them concurrently

Reply:
Yes, because it uses `await`, we can execute this concurrently